src/network/network_osm_reader.cpp

```diff
@@ -0,0 +1,361 @@
+    std::unordered_map<osmium::unsigned_object_id_type,
+      int > > node_graph;
+  
+  /** filter used to select which ways to include */
+  const way_filter wf;
+  
+  /** Function to determine if an OSM way's direction.
+   * 
+   * \returns 0 for two-way, 1 for one-way following the nodes' order,
+   *  -1 for reversed case.
+   */
+  static int is_oneway(const osmium::Way& way)  {
+		int oneway = 0;
+		const char* value = way.get_value_by_key("oneway");
+		if(value) {
+			if(!strcmp(value,"yes") || !strcmp(value,"1") || !strcmp(value,"true")) oneway = 1;
+			if(!strcmp(value,"-1")) oneway = -1;
+		}
+		if(!oneway) {
+			value = way.get_value_by_key("highway");
+			if(value && !strcmp(value,"motorway")) oneway = 1;
+			else {
+				value = way.get_value_by_key("junction");
+				if(value && !strcmp(value,"roundabout")) oneway = 1;
+			}
+		}
+		return oneway;
+	}
+  
+public:
+  /** Initialize a new instance based on the way_filter provided */
+  explicit OSMHandler(bool do_filter = false):wf(do_filter) { }
+  
+  /** Function to handle a node read from OSM data
+   * (part of the osmium::handler::Handler interface) */
+  void node(const osmium::Node& node) {
+    /* we need to store coordinates for all nodes, since we do not know
+     * which ones are in the ways to use */
+    node_coords[node.id()] = node.location();
+  }
+  
+  /** Function to handle a way read from OSM data
+   * (part of the osmium::handler::Handler interface) */
+  void way(const osmium::Way& way) {
+    /* filter out ways that are not roads */
+    if(!wf(way)) return;
+    /* check that the way has more than 1 node */
+    if(way.nodes().size() < 2) return;
+    
+    /* check the direction of this way */
+    int dir = is_oneway(way);
+    
+    /* process all nodes in this way */
+    osmium::unsigned_object_id_type last_node_id = 18446744073709551615UL;
+    for(auto& nr : way.nodes()) {
+      auto node_id = nr.ref();
+      if(last_node_id != 18446744073709551615UL) {
+        /* add all links to the graph */
+        node_graph[last_node_id][node_id] = dir;
+        node_graph[node_id][last_node_id] = -1*dir;
+      }
+      last_node_id = node_id;
+    }
+  }
+  
+  /** \brief Function that generates a simplified graph.
+   * 
+   * One edge is defined by a path between any two OSM nodes 
+   * with degree != 2. Typically, an edge contains multiple OSM 
+   * nodes which have the purpose of determining the correct 
+   * geometry of the edge. Also, there is not a one-to-one 
+   * correspondence between edges and OSM ways: an edge can 
+   * contain multiple OSM ways (e.g. if some property changes on 
+   * the way), and an OSM way can belong to multiple edge (an OSM 
+   * way can be created for a road that spans multiple 
+   * intersections).
+   * 
+   * This function identifies the edges and adds them to an
+   * FMM::NETWORK::Network class.
+   */
+  void generate_simplified_graph(Network& network) const {
```

Comment:
I think the filter is always needed but creating simplified graph should also be designed as one option. As sometimes merging the edges will lose some detailed edge information.

Reply:
I agree.

I see there could be three options:

 -- current behavior, edges are defined between OSM nodes that have degree != 2
 -- similar, but an edge can contain only one OSM way, so a graph node is kept if the way ID changes on a path even if this node has degree == 2; this way, each edge has only one OSM way associated with it, but an OSM way might correspond to multiple edges (if there is an intersection along it)
 -- OSM ways correspond directly to edges: this is the simplest case (no need to store a temporary graph), but could result in problems if a way includes an intersection in the middle

I think the current behavior is good for users who want it to work "out of the box" for any OSM network they download, but it presents the problem of relabeling edges. I think one way to deal with that is to have an option to output OSM node IDs for a path instead of edge ID, since those need not be relabeled (if using 64-bit node IDs, see below).

It could be good to give options for more advanced users who will want to perform filtering or editing themselves with external tools. I can implement the above three options that could be selected with further arguments in the constructor.

Reply:
Before you try to integrate this feature into the constructor, I am a bit concerned about it as it will make this configuration over-complicated. We want to make it both general and simple.

- I think whenever you need to change the network data, that should go to a preprocessing stage. 
- For the problems with the OSM data, if the user really needs to do such a cleaning task, then it is better to switch to creating a shapefile as input rather than using OSM file as input. That also addresses the problem saving extra output information.  
- What I expect for reading OSM file is very simple. The user just downloads an OSM file and provides it directly to the program and we can run map matching on it. That is also why I am more positive about providing filter command instructions for the user to filter out the data in need. 
- Perhaps in some regions, the network quality is much higher, like this one used http://map.project-osrm.org/ or the problem does not exist or in the future it is corrected by the users. I think it is better for the network processing functions to go to a preprocessing stage.
- That will make this program inherent and it will also be much easier to be maintained.
- I think this feature only concerns with reading OSM data. Preprocessing or cleaning road network should be treated as another feature or perhaps a separate app called `osm2network` can be provided for the user.

Reply:
I see your point. In this case, a much simplified version will be sufficient, more in line with your original implementation. In this case, I think the main change needed is to ensure 64-bit IDs are used everywhere where OSM IDs are encountered. You can of course refer people to my repository if they need to preprocess the network differently :)

Reply:
Yes, I think that is a good solution. The filtering of OSM ways is still needed here. Too complicated preprocessing such as simplifying graph can be removed here. But they can go to your program as a function.